src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClause.cs

```diff
@@ -41,29 +41,28 @@ public SelectExpandClause(IEnumerable<SelectItem> selectedItems, bool allSelecte
         }
 
         /// <summary>
-        /// Gets the selected properties and operations.
+        /// Gets a flag indicating that everything at this level has been selected.
         /// </summary>
-        /// <remarks>This list includes expanded navigations properties, which may have additional nested selections and expansions.</remarks>
-        public IEnumerable<SelectItem> SelectedItems
+        /// <remarks>
+        /// If true, then all structural properties, bound actions and functions, and all navigations in the SelectedItems list have been selected.
+        /// </remarks>
+        public bool AllSelected
         {
             get
             {
-                ////Debug.Assert(!this.usedInternalLegacyConstructor || this.selectedItems != null, "You cannot get the list of selected items until processing is complete.");
-                return new ReadOnlyCollection<SelectItem>(this.selectedItems).AsEnumerable();
+                return this.allSelected.Value;
             }
         }
 
         /// <summary>
-        /// Gets a flag indicating that everything at this level has been selected.
+        /// Gets the selected properties and operations.
         /// </summary>
-        /// <remarks>
-        /// If true, then all structural properties, bound actions and functions, and all navigations in the SelectedItems list have been selected.
-        /// </remarks>
-        public bool AllSelected
+        /// <remarks>This list includes expanded navigations properties, which may have additional nested selections and expansions.</remarks>
+        public IList<SelectItem> SelectedItems

```

Comment:
I would prefer `IReadOnlyCollection` or `IReadOnlyList`, whichever works best for your case.